### PR TITLE
feat(blocks): add Replicate model blocks

### PR DESCRIPTION
**Pull request by snehaoladri**
This PR adds two new blocks for running Replicate models in AutoGPT:

ReplicateModelBlock: Synchronous execution of any Replicate model with custom inputs

Key Features:

Support for any public Replicate model via model name (e.g., "stability-ai/stable-diffusion")
Custom input parameters via dictionary (e.g., {"prompt": "a beautiful landscape"})
Optional model version specification for reproducible results
Proper credentials handling using ProviderName.REPLICATE
Comprehensive test suite with 12 test cases covering success, error, and edge cases
Type-safe implementation with full pyright compliance
Mock methods for testing and development

# Checklist 📋
## For code changes:
- [X] I have clearly listed my changes in the PR description
- [X] I have made a test plan
- [X] I have tested my changes according to the test plan:
  - Unit tests for ReplicateModelBlock (6 test cases)
  -  Test block initialization and configuration
  -  Test mock methods for development
  -  Test error handling and edge cases
  -  Verify type safety with pyright (0 errors)
  -  Verify code formatting with Black and isort
  -  Verify linting with Ruff (0 errors)
  -  Test credentials handling with ProviderName.REPLICATE
  -  Test model version specification functionality
  -  Test both synchronous and asynchronous execution paths

### CODE REVIEW

**Comment by AutoGPT-Agent** [comment]
Thank you for your detailed PR adding Replicate model blocks to AutoGPT! The code looks well-structured with comprehensive test coverage.

### What looks good:
- Excellent test coverage with 12 test cases
- Clear documentation in the code
- Good error handling and edge case coverage
- Type-safe implementation
- Proper credential handling using ProviderName.REPLICATE

### Changes needed:

1. **PR Title Format**: Your PR title needs to follow the conventional commit format. Please update it to something like:
   ```
   feat(blocks): add Replicate model blocks
   ```
   This should specify the type (feat) and scope (blocks) according to our commit convention.

2. **Minor Suggestions**:
   - Consider adding some basic usage examples in the docstrings to make it easier for developers to use these blocks
   - The mock classes at the end of the file could be moved to the test file for cleaner separation

Once you update the PR title to follow our conventional commit format, this should be ready to merge! The implementation is solid and well-tested.

**Comment by AutoGPT-Agent** [comment]
Thanks for this comprehensive PR implementing Replicate model blocks! The code looks well-structured with strong typing, extensive error handling, and a thorough test suite.

### Required Changes

1. **PR Title Format**: Your PR title needs to follow the conventional commit format. Please update it to something like:
   ```
   feat(platform/blocks): implement replicate model blocks
   ```

### Suggestions

1. Consider adding some basic examples to the docstrings to help users understand how to use these blocks effectively.

2. The error handling is good, but consider adding more specific error messages for common API issues (rate limiting, invalid model references, etc.).

3. Your implementation correctly handles user credentials through the proper channels, which is good security practice.

The code is generally high quality and the test coverage is excellent. Once you update the PR title to match the required format, this should be ready for merging.

**Comment by snehaoladri** [comment]
@Torantulino has created a new PR addressing the issues. The Vercel deployment should be completed once the PR is reviewed and approved. Could a team member please review this so we can merge it promptly?

**Comment by ntindle** [comment]
One of our team members is testing vercel. Don’t worry about the messages from it

File: autogpt_platform/backend/backend/blocks/replicate_block.py [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+        model_name: str = SchemaField(description="Name of the model used")
+        error: str = SchemaField(description="Error message if any", default="")
+
+    @staticmethod
+    def mock_run_model(model_ref, model_inputs, api_key):
+        return "Mock response from Replicate model", "mock_prediction_id"
+
+    def __init__(self):
+        super().__init__(
+            id="a1b2c3d4-e5f6-7890-abcd-ef1234567890",
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/replicate_block.py]: please generate a real uuid :)

File: autogpt_platform/backend/backend/blocks/replicate_block.py [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+                ("model_name", str),
+            ],
+            test_mock={"run_model": staticmethod(self.mock_run_model)},
+        )
+
+    async def run(
+        self, input_data: Input, *, credentials: APIKeyCredentials, **kwargs
+    ) -> BlockOutput:
+        if isinstance(input_data, dict):
+            input_data = self.input_schema(**input_data)
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/replicate_block.py]: ?

File: autogpt_platform/backend/backend/blocks/replicate_block.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+        )
+        error: str = SchemaField(description="Error message if any", default="")
+
+    @staticmethod
+    def mock_run_async_model(model_ref, model_inputs, api_key, max_wait_time):
+        return "Mock response from Replicate model", "mock_prediction_id", 1.5
+
+    def __init__(self):
+        super().__init__(
+            id="b2c3d4e5-f6g7-8901-bcde-f23456789012",
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/replicate_block.py]: plz gen a real uuid

File: autogpt_platform/backend/backend/blocks/replicate_block.py [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+                ("execution_time", float),
+            ],
+            test_mock={"run_async_model": staticmethod(self.mock_run_async_model)},
+        )
+
+    async def run(
+        self, input_data: Input, *, credentials: APIKeyCredentials, **kwargs
+    ) -> BlockOutput:
+        if isinstance(input_data, dict):
+            input_data = self.input_schema(**input_data)
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/replicate_block.py]: ??

File: autogpt_platform/backend/backend/blocks/replicate_block.py [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+        try:
+            if input_data.version:
+                model_ref = f"{input_data.model_name}:{input_data.version}"
+            else:
+                model_ref = input_data.model_name
+            logger.info(f"Running Replicate model asynchronously: {model_ref}")
+            logger.info(f"Model inputs: {input_data.model_inputs}")
+            api_key = credentials.api_key
+            if isinstance(api_key, SecretStr):
+                api_key = api_key.get_secret_value()
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/replicate_block.py]: do not pass secrets around as strings. pass the secretstr to the run_async_model and grab the secret there

File: autogpt_platform/backend/backend/blocks/replicate_block.py [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+            print(traceback.format_exc())
+            error_msg = f"Unexpected error running Replicate model: {str(e)}"
+            logger.error(error_msg)
+            raise RuntimeError(error_msg)
+
+    async def run_async_model(
+        self,
+        model_ref: str,
+        model_inputs: Dict[str, Any],
+        api_key: str,
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/replicate_block.py]:
```suggestion
        api_key: SecretStr,
```

File: autogpt_platform/backend/backend/blocks/replicate_block_test.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
+"""
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/replicate_block_test.py]: we shouldn't need this as the block itself can have tests on it

File: autogpt_platform/backend/pyproject.toml [hunk 8]
```diff
@@ -71,6 +71,7 @@ aiofiles = "^24.1.0"
 tiktoken = "^0.9.0"
 aioclamd = "^1.0.0"
 setuptools = "^80.9.0"
+typing-extensions = "^4.14.0"
```
Comment (ntindle) [inline autogpt_platform/backend/pyproject.toml]: what is this needed for?

Comment (snehaoladri) [reply autogpt_platform/backend/pyproject.toml]: removed it

File: autogpt_platform/backend/backend/blocks/replicate_block_test.py (hunk 7 above)
Comment (snehaoladri) [reply autogpt_platform/backend/backend/blocks/replicate_block_test.py]: removed

File: autogpt_platform/backend/backend/blocks/replicate_block.py (hunk 6 above)
Comment (snehaoladri) [reply autogpt_platform/backend/backend/blocks/replicate_block.py]: done

File: autogpt_platform/backend/backend/blocks/replicate_block.py (hunk 5 above)
Comment (snehaoladri) [reply autogpt_platform/backend/backend/blocks/replicate_block.py]: done

File: autogpt_platform/backend/backend/blocks/replicate_block.py (hunk 4 above)
Comment (snehaoladri) [reply autogpt_platform/backend/backend/blocks/replicate_block.py]: Updated the logic to use execute function

File: autogpt_platform/backend/backend/blocks/replicate_block.py (hunk 3 above)
Comment (snehaoladri) [reply autogpt_platform/backend/backend/blocks/replicate_block.py]: done

File: autogpt_platform/backend/backend/blocks/replicate_block.py (hunk 2 above)
Comment (snehaoladri) [reply autogpt_platform/backend/backend/blocks/replicate_block.py]: Updated the logic

File: autogpt_platform/backend/backend/blocks/replicate_block.py (hunk 1 above)
Comment (snehaoladri) [reply autogpt_platform/backend/backend/blocks/replicate_block.py]: done

**Comment by snehaoladri** [comment]
@ntindle did you get a chance to review the PR?

**Comment by ntindle** [comment]
I'll pop this on my todo list :) waiting for some PRs to merge to fix ci then I'll take a look

**Comment by snehaoladri** [comment]
> I'll pop this on my todo list :) waiting for some PRs to merge to fix ci then I'll take a look

Sure, thank you for the update 🙂

**Comment by AutoGPT-Agent** [comment]
Thank you for this PR adding Replicate model block functionality! The implementation looks solid with good error handling and type safety.

A few observations:

1. Your PR description mentions two blocks (ReplicateModelBlock and ReplicateAsyncModelBlock), but I only see the implementation for ReplicateModelBlock in the diff. Is the ReplicateAsyncModelBlock implementation missing from the PR, or was this a change in scope?

2. The code structure looks good with proper separation of concerns - moving authentication and helper functions to separate files.

3. I appreciate the comprehensive test plan and the attention to error handling in the implementation.

4. The cost configuration for the new block has been properly added to block_cost_config.py.

Once you clarify the status of the ReplicateAsyncModelBlock (whether it's coming in a separate PR or was intended to be part of this one), this PR should be good to go.
